File: services/paper-runtime/src/unizero_runtime/contracts.py
"""Versioned HTTP contracts shared by the ZoMiner local service endpoints."""

from __future__ import annotations

from importlib.metadata import PackageNotFoundError, version
from typing import Any, Optional

from pydantic import BaseModel, ConfigDict, Field


API_VERSION = "1"
try:
    SERVICE_VERSION = version("unizero-runtime")
except PackageNotFoundError:
    # Source-tree imports before installation (for example, an editor) still get a
    # meaningful value. Installed and editable environments use pyproject metadata.
    SERVICE_VERSION = "0.1.0"


class RequestModel(BaseModel):
    model_config = ConfigDict(extra="forbid")


class ConvertOptionsPatch(RequestModel):
    backend: Optional[str] = None
    ocr_mode: Optional[str] = None
    language: Optional[str] = None
    device: Optional[str] = None
    enable_formula: Optional[bool] = None
    enable_table: Optional[bool] = None
    split_threshold: Optional[int] = None
    chunk_size: Optional[int] = None
    images_mode: Optional[str] = None
    table_mode: Optional[str] = None
    strip_repeated_lines: Optional[bool] = None
    strip_references: Optional[bool] = None
    table_vlm: Optional[bool] = None


class ConvertRequest(RequestModel):
    pdf_path: str
    title: str = ""
    authors: list[str] = Field(default_factory=list)
    year: str = ""
    doi: str = ""
    publication: str = ""
    citekey: str = ""
    item_key: str = ""
    attachment_key: str = ""
    attachment_title: str = ""
    is_supplement: bool = False
    library_id: int = 1
    library_scope: str = Field(
        default="library",
        pattern=r"^(library|groups/[1-9][0-9]*)$",
    )
    template: str = "paper-to-markdown"
    workflow: Optional[str] = None
    options: Optional[ConvertOptionsPatch] = None

    @property
    def template_id(self) -> str:
        """Prefer the new name while accepting v1's workflow field."""
        return self.workflow or self.template


class AnnotationPayload(RequestModel):
    key: str
    attachment_key: str = ""
    text: str
    comment: str = ""
    page_label: str = ""
    sort_index: str = ""


class AnnotateRequest(RequestModel):
    citekey: str = ""
    item_key: str = ""
    library_id: int = 1
    library_scope: str = Field(
        default="library",
        pattern=r"^(library|groups/[1-9][0-9]*)$",
    )
    annotations: list[AnnotationPayload] = Field(default_factory=list)


class FrontmatterConfigPatch(RequestModel):
    tags: Optional[list[str]] = None
    extra: Optional[dict[str, Any]] = None


class ConfigPatch(RequestModel):
    port: Optional[int] = None
    vault_root: Optional[str] = None
    papers_dir: Optional[str] = None
    work_dir: Optional[str] = None
    bbt_rpc: Optional[str] = None
    options: Optional[ConvertOptionsPatch] = None
    frontmatter: Optional[FrontmatterConfigPatch] = None


class JobAccepted(BaseModel):
    job_id: str
    status: str


class JobStatusResponse(BaseModel):
    job_id: str
    status: str
    log_tail: list[str] = Field(default_factory=list)
    result: Optional[dict[str, Any]] = None
    error: Optional[str] = None


class TemplateSummary(BaseModel):
    id: str
    name: str
    version: int
    description: str
    builtin: bool
    customized: bool
    stages: list[str]
    modules: list[dict[str, Any]]


class TemplateListResponse(BaseModel):
    templates: list[TemplateSummary]


class WorkflowListResponse(BaseModel):
    workflows: list[TemplateSummary]


class WorkflowTemplateDocument(BaseModel):
    schema_version: int
    id: str
    name: str
    version: int
    description: str
    modules: list[dict[str, Any]]


class TemplateDetailResponse(BaseModel):
    template: WorkflowTemplateDocument
    yaml: str
    builtin: bool
    customized: bool


class AnnotateResponse(BaseModel):
    md_path: str
    total: int
    injected: int
    already: int
    skipped: list[dict[str, Any]]
    unrouted: int


class JobListItem(BaseModel):
    job_id: str
    status: str
    title: str
    attachment_title: str = ""
    is_supplement: bool = False
    workflow: str
    template: str
    created: float
    error: Optional[str] = None
    md_path: Optional[str] = None
    last_log: str = ""


class HealthResponse(BaseModel):
    ok: bool
    service: str
    service_version: str
    api_version: str
    capabilities: list[str]
    workflows: list[dict[str, Any]]
    mineru_version: str
    vault: str
    vault_configured: bool
    papers_dir: str
    queued: int


class ConfigResponse(BaseModel):
    config: dict[str, Any]
    config_path: str


class TemplateSaveRequest(RequestModel):
    template: Optional[dict[str, Any]] = None
    yaml: Optional[str] = None


class ErrorDetail(BaseModel):
    code: str
    message: str
    details: Any = None


class ErrorResponse(BaseModel):
    error: ErrorDetail

File: apps/zotero-addon/src/runtime-client/contracts.ts
/**
 * Paper runtime HTTP contract, v1.
 *
 * 这些类型描述的是 add-on 和 Python runtime 之间那条 localhost HTTP 边界，来源是
 * ZoMiner `paper_service/api.py` 当前实现的 `/api/v1`。它们是**契约**，不是内部模型：
 * 字段名保持服务端的 snake_case，不做“顺手美化”——改一个名字就等于换一个协议版本。
 *
 * 长期归属是 packages/contracts（见 docs/PROJECT_STRUCTURE.md）。在 Phase 4 把 schema
 * 抽成语言中立定义之前，这里是 TypeScript 侧的唯一事实来源。
 */

/** 契约主版本。与服务端 `/health` 的 `api_version` 必须精确相等。 */
export const API_VERSION = "1";

export const API_PREFIX = `/api/v${API_VERSION}`;

/** 插件正常工作所必需的服务端能力；缺任何一个都直接判为不兼容。 */
export const REQUIRED_CAPABILITIES = ["convert", "annotate", "jobs"] as const;

export interface HealthResponse {
  api_version: string | number;
  capabilities?: string[];
  [key: string]: unknown;
}

export interface TemplateSummary {
  id: string;
  name?: string;
  description?: string;
  /** 用户覆盖过内置模板时为 true。 */
  overridden?: boolean;
}

export interface TemplateListResponse {
  templates?: TemplateSummary[];
}

export interface TemplateDetail extends TemplateSummary {
  workflow?: unknown;
  [key: string]: unknown;
}

/**
 * 转换请求。
 *
 * `pdf_path` 是 Zotero 附件在本机文件系统上的绝对路径——runtime 是独立进程，读不到
 * Zotero 的存储抽象，只能靠路径。这也是为什么这条边界只在 localhost 上成立。
 */
export interface ConvertRequest {
  pdf_path: string;
  attachment_key: string;
  attachment_title: string;
  is_supplement: boolean;
  library_id: number;
  template: string;
  /** 独立附件（没有父条目）时缺省。 */
  item_key?: string;
  title?: string;
  doi?: string;
  publication?: string;
  year?: string;
  authors?: string[];
  /** 由 Better BibTeX 提供，未安装时缺省。 */
  citekey?: string;
}

export interface ConvertAccepted {
  job_id: string;
}

export type JobStatus = "queued" | "running" | "done" | "failed";

export interface JobResult {
  md_path?: string;
  tables_html_path?: string;
  references?: ExtractedReference[];
  [key: string]: unknown;
}

export interface JobState {
  job_id?: string;
  status: JobStatus;
  log_tail?: string[];
  result?: JobResult;
  error?: string;
}

/**
 * 抽取工件：原始引文串 + 定位信息，不含解析后的元数据。
 * 标题/作者/年份的解析由 add-on 侧的 providers 完成，见 src/modules/zomReferences.ts。
 */
export interface ExtractedReference {
  raw?: string;
  page?: number;
  doi?: string;
  arxiv?: string;
  [key: string]: unknown;
}

export interface AnnotationPayloadItem {
  key: string;
  attachment_key: string;
  text: string;
  comment: string;
  page_label: string;
  sort_index: string;
}

export interface AnnotateRequest {
  item_key: string;
  library_id: number;
  annotations: AnnotationPayloadItem[];
  citekey?: string;
}

export interface AnnotateResponse {
  injected: number;
  total: number;
  /** 上一轮已经注入过、这次跳过的数量。注入是幂等的。 */
  already?: number;
  /** 在 Markdown 里找不到对应位置的批注。 */
  skipped?: unknown[];
}

/** 服务端统一错误体。 */
export interface RuntimeErrorBody {
  error?: {
    code?: string;
    message?: string;
  };
}
